File: net.sourceforge.reb4j.scala/src/main/scala/net/sourceforge/reb4j/scala/Expression.scala
package net.sourceforge.reb4j.scala
import java.util.regex.Pattern
import scala.util.matching.Regex


trait Expression extends Serializable with NotNull
{
	final def toPattern : Pattern = Pattern.compile(toString)
	final def toRegex(groupNames : String*) : Regex = 
		new Regex(toString, groupNames.toList : _*)
}

File: net.sourceforge.reb4j/src/main/java/net/sourceforge/reb4j/Expression.java
package net.sourceforge.reb4j;

import java.io.Serializable;
import java.util.regex.Pattern;

import fj.data.LazyString;

/**
 * Basic abstraction of a regular expression.
 * 
 * All implementations of this can be converted directly into a
 * {@link Pattern} object without risk of throwing an exception.
 * 
 * This interface is not intended to be implemented by clients,
 * and doing so will result in unpredictable behavior.
 */
public interface Expression extends Serializable 
{
	/**
	 * Returns the regular expression represented by this object, in a form 
	 * suitable to passing to the {@link java.util.regex.Pattern} class.
	 */
	LazyString expression();
	
	/**
	 * Passes the regular expression represented by this object to 
	 * {@link java.util.regex.Pattern} and returns the result.
	 */
	Pattern toPattern();
}
